File: apps/website/lib/financial-onboarding.ts
// Financial onboarding for a Developer Project — the pure half.
//
// "One Business identity. Multiple onboarding surfaces. One KYB authority."
// A Project gets a Business to receive into in exactly two ways, both inside the
// Developers Console:
//
//   A. it applies for a NEW Business — the same application the public form
//      collects, reviewed by an operator in BANZADMIN; approval provisions the
//      Business (@banza, wallet) and binds the Project to it;
//   B. it connects an EXISTING Banzami Business, with that Business's consent:
//      a single-use code the Business issues from its own app.
//
// The one-click setup that created a synthetic, self-approved Business is
// retired (the server answers it 410 FINANCIAL_SETUP_BY_REVIEW) and nothing here
// can reach it.
//
// Everything in this module is framework-free so it can be tested without a DOM:
// which state a Project is in, what each blocker means, how a consent code is
// typed, and what each refusal says.

import type { FinancialSetupState, OnboardingApplication, OnboardingState } from './developer-api';

// ── State ──────────────────────────────────────────────────────────────────

/** The states the Financial Setup page renders: the onboarding states, plus
 *  UNAVAILABLE for a deployment that cannot onboard at all. */
export type OnboardingView = OnboardingState | 'UNAVAILABLE';

const ONBOARDING_STATES: OnboardingState[] = [
  'NOT_CONFIGURED', 'IN_REVIEW', 'INFORMATION_REQUIRED', 'APPROVED_PROVISIONING', 'REJECTED', 'READY', 'BLOCKED',
];

/**
 * Which state to render for a Project.
 *
 * The server's onboarding state is the answer, with two exceptions. A
 * deployment that cannot onboard (financial setup UNAVAILABLE) is shown as
 * that, whatever the role allows — "you may not" and "nobody can here" lead to
 * different places. And a server that predates onboarding is read from the
 * financial setup state it does send, never guessed at.
 */
export function onboardingViewOf(setup: FinancialSetupState): OnboardingView {
  if (setup.state === 'UNAVAILABLE') return 'UNAVAILABLE';
  const s = setup.onboarding?.state;
  if (s && ONBOARDING_STATES.includes(s)) return s;
  if (setup.state === 'READY' || setup.state === 'SEALED') {
    return (setup.readiness?.settlement.blockers.length ?? 0) > 0 ? 'BLOCKED' : 'READY';
  }
  return 'NOT_CONFIGURED';
}

/** Short status label, as shown after "Configuração financeira —". */
export const ONBOARDING_LABEL: Record<OnboardingView, string> = {
  NOT_CONFIGURED: 'Não configurado',
  IN_REVIEW: 'Em análise',
  INFORMATION_REQUIRED: 'Informação pedida',
  APPROVED_PROVISIONING: 'Aprovado — a concluir a configuração',
  REJECTED: 'Candidatura recusada',
  READY: 'Pronto',
  BLOCKED: 'Bloqueado',
  UNAVAILABLE: 'Indisponível',
};

/** The page heading for each state. */
export function onboardingHeading(view: OnboardingView): string {
  // Approval is the news; the heading says it on its own.
  if (view === 'APPROVED_PROVISIONING') return ONBOARDING_LABEL.APPROVED_PROVISIONING;
  return `Configuração financeira — ${ONBOARDING_LABEL[view]}`;
}

/**
 * Whether this member may start an application, connect a Business or answer
 * a reviewer. Advice to the UI only: the server authorises every attempt.
 */
export function canActOn(setup: FinancialSetupState): boolean {
  if (setup.state === 'UNAVAILABLE') return false;
  if (setup.onboarding) return setup.onboarding.can_act;
  return setup.can_configure;
}

/** OWNER and ADMIN — the same line the server draws for financial setup. */
export function isFinancialActor(role: string): boolean {
  return role === 'OWNER' || role === 'ADMIN';
}

/** The application reference a developer quotes to support: 8 characters. */
export function applicationReference(applicationId: string): string {
  return applicationId.replace(/-/g, '').slice(0, 8).toUpperCase();
}

// ── Readiness blockers ─────────────────────────────────────────────────────

/**
 * What each readiness blocker means, in the developer's terms. The codes are
 * the refusals settlement returns (core settlement_readiness.rs); an unknown
 * one is still shown, by its code, and still blocks.
 *
 * FEE_DESTINATION_TYPE_NOT_ALLOWED is NOT a verification problem: the Business
 * is verified, and what is missing is the operator's classification of the
 * account as one that may receive application fees.
 */
export const BLOCKER_TEXT: Record<string, string> = {
  FINANCIAL_SETUP_NOT_CONFIGURED: 'Este projeto ainda não está ligado a um negócio verificado.',
  WALLET_MISSING: 'O negócio não tem uma carteira activa em Kwanza.',
  PRICING_NOT_CONFIGURED: 'O Banzami ainda não atribuiu um preço a este projeto. Contacte o suporte.',
  PRICING_CONFIGURATION_ERROR: 'O preço atribuído a este projeto está mal configurado. Contacte o suporte.',
  FEE_DESTINATION_NOT_FOUND: 'O destino das taxas de aplicação não foi encontrado: não tem @banza ou conta associada.',
  FEE_DESTINATION_NOT_OWNED: 'O destino das taxas de aplicação não pertence a este projeto.',
  FEE_DESTINATION_NOT_BUSINESS_ACCOUNT: 'O destino das taxas de aplicação não é uma conta Banzami Business.',
  FEE_DESTINATION_NOT_ACTIVE: 'A conta de destino das taxas de aplicação não está activa.',
  FEE_DESTINATION_KYB_NOT_APPROVED: 'A verificação (KYB) do destino das taxas de aplicação não está aprovada.',
  FEE_DESTINATION_WALLET_UNAVAILABLE: 'A carteira do destino das taxas de aplicação não está activa.',
  FEE_DESTINATION_TYPE_NOT_ALLOWED:
    'Receber taxas de aplicação requer aprovação do operador Banzami (classificação da conta).',
  WEBHOOK_ENDPOINT_MISSING:
    'Sem endpoint de webhook: a sua aplicação só saberá que uma liquidação terminou se a consultar.',
};

export function blockerText(code: string): string {
  return BLOCKER_TEXT[code] ?? code;
}

// ── Consent code (path B) ──────────────────────────────────────────────────

/** The alphabet the Business App draws codes from: no 0/O, 1/I/L. */
const LINK_CODE_ALPHABET = /^[ABCDEFGHJKMNPQRSTUVWXYZ23456789]{12}$/;

/** What a person typed, as the server reads it: upper case, no dashes or spaces. */
export function normaliseLinkCode(raw: string): string {
  return raw.toUpperCase().replace(/[^A-Z0-9]/g, '').slice(0, 12);
}

/** Grouped for reading, as the Business App shows it: ABCD-EFGH-JKMN. */
export function formatLinkCode(raw: string): string {
  const c = normaliseLinkCode(raw);
  return (c.match(/.{1,4}/g) ?? []).join('-');
}

/** Twelve characters of the code alphabet. Anything else cannot be a code. */
export function isCompleteLinkCode(raw: string): boolean {
  return LINK_CODE_ALPHABET.test(normaliseLinkCode(raw));
}

// ── Refusals ───────────────────────────────────────────────────────────────

/**
 * What a refusal from either onboarding path means. Codes come through from
 * the Gateway unchanged (developer-api onboardingErr); an unknown one gets a
 * sentence that sends the developer to try again, never a raw English message.
 */
export const REFUSAL_TEXT: Record<string, string> = {
  INVALID_HANDLE: 'O @banza tem de ter 3 a 30 caracteres: letras minúsculas, números ou _.',
  VALIDATION_ERROR: 'Alguns dados não foram aceites. Reveja o formulário e tente novamente.',
  VALIDATION: 'Alguns dados não foram aceites. Reveja o formulário e tente novamente.',
  APPLICATION_IN_PROGRESS:
    'Este projeto já tem uma candidatura em curso. Aguarde a decisão do operador antes de ligar ou pedir outro negócio.',
  HANDLE_RESERVED: 'Este @banza está reservado. Escolha outro.',
  HANDLE_OWNED_BY_BUSINESS:
    'Este @banza já é de um negócio Banzami. Peça ao negócio um código e ligue-o a este projeto, em vez de criar outro.',
  HANDLE_TAKEN: 'Este @banza já não está disponível. Escolha outro.',
  PROJECT_ALREADY_RECEIVING: 'Este projeto já recebe num negócio. Mudar de negócio é uma decisão do operador.',
  ONBOARDING_UNAVAILABLE: 'A verificação de negócios não está disponível agora. Tente novamente mais tarde.',
  FORBIDDEN: 'Só um Owner ou Admin do workspace pode fazer isto.',
  LINK_CODE_INVALID:
    'O código não é válido ou já expirou. Peça ao negócio um código novo — vale 10 minutos e só pode ser usado uma vez.',
  BUSINESS_NOT_READY:
    'Este negócio ainda não pode receber pagamentos, por isso não pode ser ligado: tem de estar activo, com @banza e carteira em Kwanza.',
  RATE_LIMITED: 'Demasiados pedidos. Tente novamente daqui a pouco.',
  UNAUTHENTICATED: 'A sua sessão expirou. Inicie sessão novamente.',
  NETWORK: 'Sem ligação ao serviço. Tente novamente.',
};

export function refusalText(code: string | undefined): string {
  return (code && REFUSAL_TEXT[code]) || 'Não foi possível concluir agora. Tente novamente.';
}

/**
 * Why a requested @banza is not available (the public check-handle reasons).
 * BUSINESS is the one with somewhere to go: the @ is an existing Business's,
 * and the way to use it is that Business's consent code, not a new application.
 */
export function handleUnavailableText(reason: string | undefined): string {
  switch (reason) {
    case 'BUSINESS':
      return 'Este @ já é de um negócio — ligue-o com o código do negócio.';
    case 'TAKEN':
      return 'Este @banza já está em uso.';
    case 'RESERVED':
      return 'Este @banza está reservado.';
    case 'PENDING':
      return 'Este @banza já tem uma candidatura em curso.';
    case 'INVALID':
      return 'Use 3 a 30 caracteres: letras minúsculas, números ou _.';
    default:
      return 'Este @banza não está disponível.';
  }
}

// ── An application in review ───────────────────────────────────────────────

/**
 * One requirement's standing, in words. The Gateway's reasons are MISSING,
 * UPLOADED, ACCEPTED and REJECTED[: why]; the reviewer's request for
 * information arrives as the reviewer's own words and is shown as they are.
 */
export function requirementReasonText(reason: string): string {
  if (reason === 'MISSING') return 'Em falta';
  if (reason === 'UPLOADED') return 'Enviado · a aguardar verificação';
  if (reason === 'ACCEPTED') return 'Aceite';
  if (reason === 'REJECTED') return 'Recusado';
  if (reason.startsWith('REJECTED: ')) return `Recusado: ${reason.slice('REJECTED: '.length)}`;
  return reason;
}

/** The code the Gateway gives the reviewer's request inside `errors`. It is
 *  shown on its own, prominently, not as one more line of the list. */
export const INFORMATION_REQUEST_CODE = 'information_request';

const UPLOADABLE_TYPES = ['BUSINESS_REGISTRATION', 'REPRESENTATIVE_ID', 'TAX_ID', 'OTHER'] as const;
export type UploadableDocument = (typeof UPLOADABLE_TYPES)[number];

/**
 * Which documents may be sent for an application now: every document the
 * policy still wants (missing) or that the reviewer refused (a replacement),
 * plus the optional additional document. A document awaiting verification or
 * already accepted is not offered again.
 */
export function uploadableDocuments(app: OnboardingApplication): UploadableDocument[] {
  const wanted = [...app.requirements.currently_due, ...app.requirements.errors]
    .filter((r) => r.kind === 'document')
    .map((r) => r.code)
    .filter((c): c is UploadableDocument => (UPLOADABLE_TYPES as readonly string[]).includes(c));
  const out = [...new Set(wanted)];
  if (!out.includes('OTHER')) out.push('OTHER');
  return out;
}

/** What a refused resubmission means. */
export function resubmitRefusalText(code: string | undefined): string {
  switch (code) {
    case 'REQUIREMENTS_NOT_MET':
      return 'Ainda há requisitos por cumprir. Envie os documentos em falta ou recusados e tente novamente.';
    case 'NOT_WAITING_FOR_INFORMATION':
      return 'A candidatura já não está à espera de informação.';
    case 'APPLICATION_NOT_FOUND':
      return 'Não encontrámos esta candidatura. Atualize a página.';
    default:
      return refusalText(code);
  }
}

// ── The application form ───────────────────────────────────────────────────

/**
 * The fields the requirements policy requires, as of ao-business-2026-09.
 *
 * Used only when the policy cannot be read. The Gateway is the authority — it
 * refuses an approval whose requirements are not met, whatever this form said —
 * so the form fetches GET /v1/merchant/application-requirements and mirrors it,
 * and falls back to this copy only so that an unreachable policy does not leave
 * a form that requires nothing.
 */
export const FALLBACK_REQUIRED_FIELDS = [
  'business_name', 'desired_handle', 'category', 'email', 'phone', 'nif', 'province', 'municipality',
  'address', 'legal_representative', 'representative_role', 'business_activity', 'terms_accepted',
] as const;

export const FALLBACK_REQUIRED_DOCUMENTS = ['BUSINESS_REGISTRATION', 'REPRESENTATIVE_ID'] as const;

/** Required field and document codes from a policy (or the fallback). */
export function requiredFrom(
  policy: { items: { code: string; kind: string }[] } | null,
): { fields: Set<string>; documents: Set<string> } {
  if (!policy || policy.items.length === 0) {
    return { fields: new Set(FALLBACK_REQUIRED_FIELDS), documents: new Set(FALLBACK_REQUIRED_DOCUMENTS) };
  }
  return {
    fields: new Set(policy.items.filter((i) => i.kind === 'field').map((i) => i.code)),
    documents: new Set(policy.items.filter((i) => i.kind === 'document').map((i) => i.code)),
  };
}

/** The roles a legal representative can hold — the public form's list. */
export const REPRESENTATIVE_ROLES = ['Proprietário(a)', 'Sócio(a)', 'Gerente', 'Administrador(a)', 'Representante legal'];

/** The three document slots: two the policy requires, one optional. */
export const DOCUMENT_SLOTS: { type: 'BUSINESS_REGISTRATION' | 'REPRESENTATIVE_ID' | 'OTHER'; label: string; hint: string }[] = [
  { type: 'BUSINESS_REGISTRATION', label: 'Registo Comercial', hint: 'Certidão do registo comercial da empresa' },
  { type: 'REPRESENTATIVE_ID', label: 'Documento de identidade do representante', hint: 'BI ou Passaporte' },
  { type: 'OTHER', label: 'Documento adicional', hint: 'Licença, declaração ou outro documento relevante' },
];

export function documentLabel(type: string): string {
  return DOCUMENT_SLOTS.find((d) => d.type === type)?.label ?? type;
}

export const MAX_DOCUMENT_BYTES = 5 * 1024 * 1024;
export const DOCUMENT_ACCEPT = '.pdf,.jpg,.jpeg,.png,application/pdf,image/jpeg,image/png';

/** PDF, JPEG or PNG, at most 5 MB. Null when acceptable. */
export function documentProblem(file: { name: string; type: string; size: number }): string | null {
  const okType = ['application/pdf', 'image/jpeg', 'image/png'].includes(file.type) || /\.(pdf|jpe?g|png)$/i.test(file.name);
  if (!okType) return 'Formato inválido. Use PDF, JPEG ou PNG.';
  if (file.size > MAX_DOCUMENT_BYTES) return 'Ficheiro demasiado grande (máximo 5 MB).';
  if (file.size === 0) return 'O ficheiro está vazio.';
  return null;
}

/** What the document upload says when the deployment has no document storage. */
export const STORAGE_NOT_CONFIGURED_TEXT = 'O envio de documentos ainda não está disponível neste ambiente.';

export const isEmail = (s: string) => /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(s.trim());
/** An Angolan number: 9 digits after +244. */
export const isAngolanPhone = (s: string) => /^\d{9}$/.test(s.replace(/\D/g, ''));
/** NIF: digits only, 9 to 14 of them. */
export const isNif = (s: string) => /^\d{9,14}$/.test(s.replace(/\D/g, ''));

/** A fresh idempotency key — one per form session, never per attempt. */
export function newIdempotencyKey(): string {
  if (typeof crypto !== 'undefined' && 'randomUUID' in crypto) return crypto.randomUUID();
  return `fo-${Date.now()}-${Math.random().toString(36).slice(2)}`;
}

File: services/developer-api/internal/developer/financial_onboarding.go
package developer

import (
	"context"
	"errors"
	"log/slog"
	"strings"

	"github.com/banzami/banzami/services/developer-api/internal/gatewayclient"
)

// Financial onboarding — how a Project becomes able to receive money.
//
// One Business identity, several onboarding surfaces, one KYB authority. A
// Project does not get a Business of its own making: it either
//
//   - applies for a NEW Business through the same Business application the
//     public form submits — same fields, documents and operator review in
//     BANZADMIN — and approval provisions the Business and binds the Project to
//     it (the Gateway records that as a provisioning step); or
//   - connects a Business that ALREADY EXISTS, with that Business's consent: the
//     Business App issues a single-use code, and redeeming it here binds the
//     Project to it (ADR-055 binding). Nothing is re-verified or re-created, and
//     no handle, wallet or ledger history moves.
//
// This replaces the one-click Sandbox setup that created a synthetic Business
// and marked its KYB approved with nobody reviewing anything. A Project can
// still exist, take keys and integrate without any of this; it only needs a
// Business to receive.

// Onboarding states a Project is shown. About the Project, not the tables.
const (
	OnboardingNotConfigured        = "NOT_CONFIGURED"
	OnboardingInReview             = "IN_REVIEW"
	OnboardingInformationRequired  = "INFORMATION_REQUIRED"
	OnboardingApprovedProvisioning = "APPROVED_PROVISIONING"
	OnboardingRejected             = "REJECTED"
	OnboardingReady                = "READY"
	OnboardingBlocked              = "BLOCKED"
	// OnboardingReadinessUnknown: the Project is bound to a Business, but
	// whether it can settle could not be read just now. Never shown as READY:
	// a failed read is not an answer (A2-26).
	OnboardingReadinessUnknown = "READINESS_UNKNOWN"
)

var (
	// ErrOnboardingUnavailable: this deployment cannot reach the onboarding
	// domain.
	ErrOnboardingUnavailable = errors.New("financial onboarding is not available on this deployment")
	// ErrProjectAlreadyReceiving: the Project already has a Business; a second
	// one is a rebind, which is an operator decision (and refused once sealed).
	ErrProjectAlreadyReceiving = errors.New("this Project already receives into a Business")
	// ErrOneClickSetupRetired: the synthetic self-approved Sandbox Business is
	// gone; a Project applies or connects a Business.
	ErrOneClickSetupRetired = errors.New("a Project's Business is applied for and reviewed, or connected with the Business's consent")
)

// BusinessOnboarding is the Gateway's Business application domain.
type BusinessOnboarding interface {
	LatestForProject(ctx context.Context, projectID string) (*gatewayclient.ProjectApplication, error)
	SubmitForProject(ctx context.Context, in gatewayclient.ApplicationInput) (string, error)
	RedeemLinkCode(ctx context.Context, code, projectID string) (*gatewayclient.LinkTarget, error)
}

// BusinessNamer reads a Business's display name.
type BusinessNamer interface {
	MerchantName(ctx context.Context, merchantID string) (string, error)
}

// BusinessIdentities reads a Business's PUBLIC identity — the name it presents
// and the @banza it owns (the Gateway's business_public_identities). The
// Gateway onboarding client implements it.
type BusinessIdentities interface {
	BusinessPublicIdentity(ctx context.Context, merchantID string) (*gatewayclient.BusinessIdentity, error)
}

// businessPublicIdentity is the Business a Project receives into, as the Console
// names it. The Business's public identity, not merchants.name: for a Business
// made by the retired Console setup the account name is "Sandbox · <Project>",
// and the card under "Este projeto recebe pagamentos no negócio abaixo" showed
// exactly that. Only when no identity source is wired does the account name
// stand in; a failed lookup names nobody.
func (s *Service) businessPublicIdentity(ctx context.Context, merchantID string) (gatewayclient.BusinessIdentity, bool) {
	if ids, ok := s.onboarding.(BusinessIdentities); ok && ids != nil {
		b, err := ids.BusinessPublicIdentity(ctx, merchantID)
		if err != nil || b == nil {
			slog.WarnContext(ctx, "developer.financial_onboarding.business_identity_unavailable", "err", errString(err))
			return gatewayclient.BusinessIdentity{}, false
		}
		return *b, true
	}
	if s.namer != nil {
		if n, err := s.namer.MerchantName(ctx, merchantID); err == nil {
			return gatewayclient.BusinessIdentity{DisplayName: n}, true
		}
	}
	return gatewayclient.BusinessIdentity{}, false
}

func errString(err error) string {
	if err == nil {
		return "no identity"
	}
	return err.Error()
}

// SetBusinessOnboarding wires the Gateway's onboarding domain.
func (s *Service) SetBusinessOnboarding(o BusinessOnboarding) { s.onboarding = o }

// SetBusinessNamer wires the Business name lookup for a bound Project.
func (s *Service) SetBusinessNamer(n BusinessNamer) { s.namer = n }

// OnboardingBusiness is the Business a Project receives into, as the Console
// shows it: its public identity and whether it is verified. No internal id.
type OnboardingBusiness struct {
	Name      string `json:"name"`
	Handle    string `json:"handle"`
	KybStatus string `json:"kyb_status"`
	Verified  bool   `json:"verified"`
}

// FinancialOnboarding is the Project's onboarding, one level above the tables.
type FinancialOnboarding struct {
	State string `json:"state"`
	// CanAct: this member may start an application or connect a Business.
	CanAct      bool                              `json:"can_act"`
	Business    *OnboardingBusiness               `json:"business,omitempty"`
	Application *gatewayclient.ProjectApplication `json:"application,omitempty"`
	// Blockers are the reasons a READY-looking Project cannot settle yet —
	// settlement readiness's own codes, for the Console to explain.
	Blockers []string `json:"blockers"`
}

// onboardingView derives the Project's onboarding state. Called with the
// Project's binding (nil when unbound), its readiness when bound, and whether
// the readiness read failed.
func (s *Service) onboardingView(ctx context.Context, projectID, role string, b *SandboxBinding, r *ProjectReadiness, readinessUnavailable bool) *FinancialOnboarding {
	v := &FinancialOnboarding{State: OnboardingNotConfigured, CanAct: canConfigureFinancialSandbox(role), Blockers: []string{}}
	if b != nil && b.MerchantID != "" {
		v.CanAct = false
		bus := &OnboardingBusiness{}
		identity, known := s.businessPublicIdentity(ctx, b.MerchantID)
		if known {
			bus.Name = identity.DisplayName
			if h := strings.TrimPrefix(identity.Handle, "@"); h != "" {
				bus.Handle = "@" + h
			}
		}
		if r != nil {
			if bus.Handle == "" && r.FinancialIdentity.Handle != nil {
				bus.Handle = *r.FinancialIdentity.Handle
			}
			if r.Kyb.Status != nil {
				bus.KybStatus = *r.Kyb.Status
			}
			bus.Verified = bus.KybStatus == "APPROVED"
			v.Blockers = append(v.Blockers, r.Settlement.Blockers...)
		}
		v.Business = bus
		if readinessUnavailable {
			// No readiness, no blockers — which read as READY. Whether this
			// Project can settle is simply not known right now (A2-26).
			v.State = OnboardingReadinessUnknown
			return v
		}
		v.State = OnboardingReady
		if len(v.Blockers) > 0 {
			v.State = OnboardingBlocked
		}
		return v
	}
	if s.onboarding == nil {
		return v
	}
	app, err := s.onboarding.LatestForProject(ctx, projectID)
	if err != nil {
		slog.WarnContext(ctx, "developer.financial_onboarding.application_unavailable", "project", projectID, "err", err.Error())
		return v
	}
	if app == nil {
		return v
	}
	v.Application = app
	switch app.Status {
	case "SUBMITTED", "UNDER_REVIEW", "DRAFT":
		v.State, v.CanAct = OnboardingInReview, false
	case "INFORMATION_REQUIRED":
		v.State = OnboardingInformationRequired
	case "APPROVED", "PROVISIONING_FAILED":
		// Approved, and the Project is not bound yet: the binding is the last
		// provisioning step, retried by the operator.
		v.State, v.CanAct = OnboardingApprovedProvisioning, false
	case "REJECTED", "CANCELLED":
		v.State = OnboardingRejected
	}
	return v
}

// SubmitFinancialApplication starts a Project's application for a new
// Business. The Project and the submitting member come from the session; the
// browser supplies only the Business's details.
func (s *Service) SubmitFinancialApplication(ctx context.Context, actor, projectID string, in gatewayclient.ApplicationInput, ip, reqID string) (string, error) {
	p, role, err := s.projectAuthz(ctx, actor, projectID)
	if err != nil {
		return "", err
	}
	if !canConfigureFinancialSandbox(role) {
		return "", ErrForbidden
	}
	if s.onboarding == nil {
		return "", ErrOnboardingUnavailable
	}
	if b, err := s.store.ActiveBindingForProject(ctx, p.ID); err != nil {
		return "", ErrUnavailable
	} else if b != nil {
		return "", ErrProjectAlreadyReceiving
	}
	in.ProjectID, in.SubmittedByUserID = p.ID, actor
	id, err := s.onboarding.SubmitForProject(ctx, in)
	if err != nil {
		return "", err
	}
	s.audit(ctx, &actor, &p.WorkspaceID, &projectID, "project.financial_application_submitted",
		"APPLICATION:"+id, ip, reqID, map[string]any{"requested_handle": strings.TrimSpace(in.DesiredHandle)})
	return id, nil
}

// LinkExistingBusiness connects a Business that already exists to the
// Project, with the Business's consent code. The Business is not re-verified
// or re-created; the Project is bound to it (ADR-055) and receives into it.
func (s *Service) LinkExistingBusiness(ctx context.Context, actor, projectID, code, ip, reqID string) (*OnboardingBusiness, error) {
	p, role, err := s.projectAuthz(ctx, actor, projectID)
	if err != nil {
		return nil, err
	}
	if !canConfigureFinancialSandbox(role) {
		return nil, ErrForbidden
	}
	if s.onboarding == nil {
		return nil, ErrOnboardingUnavailable
	}
	if b, err := s.store.ActiveBindingForProject(ctx, p.ID); err != nil {
		return nil, ErrUnavailable
	} else if b != nil {
		return nil, ErrProjectAlreadyReceiving
	}
	// The guard below is the only thing that stops a Project from being bound
	// here while its own application would provision a second Business. It was
	// skipped whenever the Gateway could not answer; a guard that cannot be
	// checked now refuses (A2-25).
	app, err := s.onboarding.LatestForProject(ctx, p.ID)
	if err != nil {
		slog.WarnContext(ctx, "developer.financial_onboarding.application_unavailable", "project", projectID, "err", err.Error())
		return nil, ErrOnboardingUnavailable
	}
	if app != nil {
		switch app.Status {
		case "SUBMITTED", "UNDER_REVIEW", "INFORMATION_REQUIRED", "PROVISIONING_FAILED", "APPROVED":
			// An application in progress would provision a second Business for
			// the same Project. One or the other.
			return nil, &gatewayclient.Refusal{Status: 409, Code: "APPLICATION_IN_PROGRESS",
				Message: "this Project has an application in progress"}
		}
	}
	t, err := s.onboarding.RedeemLinkCode(ctx, code, p.ID)
	if err != nil {
		return nil, err
	}
	if _, err := s.BindProjectSandbox(ctx, p.ID, t.MerchantID, t.WalletID, t.WalletAccountID, actor, ip, reqID); err != nil {
		return nil, err
	}
	s.audit(ctx, &actor, &p.WorkspaceID, &projectID, "project.business_linked",
		"PROJECT:"+projectID, ip, reqID, map[string]any{"merchant_id": t.MerchantID, "handle": t.Handle})
	// Named by its public identity, as the Project's card will name it on every
	// later read; the redeem answer carries the account name.
	name := t.BusinessName
	if _, ok := s.onboarding.(BusinessIdentities); ok {
		identity, known := s.businessPublicIdentity(ctx, t.MerchantID)
		name = identity.DisplayName
		if !known {
			name = ""
		}
	}
	return &OnboardingBusiness{Name: name, Handle: "@" + strings.TrimPrefix(t.Handle, "@"), KybStatus: t.KybStatus, Verified: t.KybStatus == "APPROVED"}, nil
}
